File: solutions/go/circular-buffer/1/circular_buffer.go
package circular

import "errors"

type Buffer struct{
    buf		[]byte
    len		int
    cap		int
    start	int
    end		int
}

func NewBuffer(size int) *Buffer {
    return &Buffer{
        buf:	make([]byte, size),
        len:	0,
        cap:	size,
        start:	0,
        end:	0,
    }
}

func (b *Buffer) ReadByte() (r byte, e error) {
    if b.len == 0 {
        e = errors.New("Read from empty buffer!")
        return
    }
    r = b.buf[b.start]
    b.start = (b.start+1) % b.cap
    b.len--
    return
}

func (b *Buffer) WriteByte(c byte) (e error) {
    if b.len == b.cap {
        e = errors.New("Write on full buffer!")
        return
    }
    b.buf[b.end] = c
    b.end = (b.end+1) % b.cap
    b.len++
    return
}

func (b *Buffer) Overwrite(c byte) {
    if b.len == b.cap {
        b.ReadByte()
    }
    b.WriteByte(c)
    return
}

func (b *Buffer) Reset() {
    b.start = 0
    b.end = 0
    b.len = 0
    return
}

File: solutions/typescript/circular-buffer/1/circular-buffer.ts
class EmptyMarker {}

export default class CircularBuffer<T> {
  private buffer: Array<T | EmptyMarker>
  private readCursor: number = 0
  private writeCursor: number = 0
  private static EMPTY = new EmptyMarker()
  
  constructor(initialCapacity: number) {
    this.buffer = Array.from(
      { length: Math.max(1, Math.floor(initialCapacity)) },
      () => CircularBuffer.EMPTY
    )
  }

  write(value: T): void | never {
    const oldValue = this.buffer[this.writeCursor]
    if (oldValue instanceof EmptyMarker) {
      this.buffer[this.writeCursor] = value
      this.writeCursor = (this.writeCursor + 1) % this.buffer.length
    } else {
      throw new BufferFullError()
    }
  }

  read(): T | never {
    const value = this.buffer[this.readCursor]
    if (value instanceof EmptyMarker) {
      throw new BufferEmptyError()
    } else {
      this.buffer[this.readCursor] = CircularBuffer.EMPTY
      this.readCursor = (this.readCursor + 1) % this.buffer.length
      return value
    }
  }

  forceWrite(value: T): void {
    const oldValue = this.buffer[this.writeCursor]
    if (oldValue instanceof EmptyMarker) {
      this.buffer[this.writeCursor] = value
      this.writeCursor = (this.writeCursor + 1) % this.buffer.length
    } else {
      this.read()
      this.write(value)
    }
  }

  clear(): void {
    this.buffer.fill(CircularBuffer.EMPTY)
    this.readCursor = 0
    this.writeCursor = 0
  }
}

export class BufferFullError extends Error {
  constructor(message?: string) {
    super(message || 'Buffer is full')
  }
}

export class BufferEmptyError extends Error {
  constructor(message?: string) {
    super(message || 'Buffer is empty')
  }
}